### ISSUE TITLE
azurerm_mssql_virtual_machine auto_backup encryption_password failing

### ISSUE THREAD

**Issue by dkisby** [hashicorp/terraform-provider-azurerm#12811]
### Terraform (and Provider) Versions
Terraform Version: v1.0.3
azurerm v2.70.0
random v3.1.0

### Affected resources
`azurerm_mssql_virtual_machine
`
### Terraform configuration files
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "rg" {
  name     = "test-rg"
  location = "West Europe"
}

resource "azurerm_network_interface" "nic" {
  name                = "nic-vmtest"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name

  ip_configuration {
    name                          = "internal"
    subnet_id                     = "<subnetID>"
    private_ip_address_allocation = "Dynamic"
  }
}

resource "azurerm_windows_virtual_machine" "vm" {
  name                = "testvm"
  resource_group_name = azurerm_resource_group.rg.name
  location            = azurerm_resource_group.rg.location
  size                = "Standard_D2as_v4"
  admin_username      = "adminuser"
  admin_password      = "P@$$w0rd1234!"
  network_interface_ids = [
    azurerm_network_interface.nic.id,
  ]

  os_disk {
    caching              = "ReadWrite"
    storage_account_type = "Standard_LRS"
  }

  source_image_reference {
    publisher = "MicrosoftSQLServer"
    offer     = "sql2016sp2-ws2019"
    sku       = "standard"
    version   = "latest"
  }
}

resource "azurerm_managed_disk" "datadisk" {
    count                = 2
    name                 = "DataDisk_${count.index +1}"
    location             = azurerm_resource_group.rg.location
    resource_group_name  = azurerm_resource_group.rg.name
    storage_account_type = "Standard_LRS"
    create_option        = "Empty"
    disk_size_gb         = 128
}

resource "azurerm_virtual_machine_data_disk_attachment" "datadisk01" {
    count              = 2
    managed_disk_id    = element(azurerm_managed_disk.datadisk.*.id, count.index)
    virtual_machine_id = element(azurerm_windows_virtual_machine.vm.*.id, count.index)
    lun                = "1${count.index}"
    caching            = "ReadWrite"
}

resource "random_password" "encryptionpassword" {
    length = 12
    special = true
}

resource "azurerm_mssql_virtual_machine" "db_vm" {
  virtual_machine_id               = azurerm_windows_virtual_machine.vm.id
  sql_license_type                 = "PAYG"
  auto_patching {
    day_of_week                            = "Saturday"
    maintenance_window_duration_in_minutes = 60
    maintenance_window_starting_hour       = 6
  }
  auto_backup {
    encryption_enabled              = true
    encryption_password             = random_password.encryptionpassword.result
    retention_period_in_days        = 14
    system_databases_backup_enabled = true
    storage_blob_endpoint           = "https://test.blob.core.windows.net/automaticbackup"
    storage_account_access_key      = "<accessKey>"
    manual_schedule {
      full_backup_frequency           = "Daily"
      full_backup_start_hour          = 4
      full_backup_window_in_hours     = 4
      log_backup_frequency_in_minutes = 60
    }
  }
}
```
### Expected behaviour
Encryption password is created and used as the encryption_password in auto_backup

### Actual behaviour
 ```
terraform plan
Error: auto_backup: `encryption_password` is required when `encryption_enabled` is true
 with azurerm_mssql_virtual_machine.db_vm,
  on main.tf line 69, in resource "azurerm_mssql_virtual_machine" "db_vm":
  69: resource "azurerm_mssql_virtual_machine" "db_vm" {
```
### Steps to Reproduce

1. terraform plan

**Comment by wasfree** [comment]
Hi @dkisby,

seems this issue is related to a [pre-checking](https://github.com/terraform-providers/terraform-provider-azurerm/blob/50a50b3db89aa71b979c669d1e78bddcf737962d/azurerm/internal/services/mssql/mssql_virtual_machine_resource.go#L308) if `encryption_enabled` is true and `encryption_password` has a empty string. Unfortunately the `random_password` resource will provide the result only after apply, while the pre-checking expecting a non-empty string on a earlier stage. 

There are few ways to resolve this issue:
1. Remove this pre-checking and rely on documentation. So it's up to the user to provider a proper configuration.
2. We could deprecate `encryption_enabled` and only look for `encryption_password`. If `encryption_password` was set we also enable encryption, if not we disable it. This would also make the pre-checking obsolete.

To be honest `encryption_enabled` isn't really needed at all, if someone want to encrypt his backup and provided a `encryption_password` this should be enough. As a third way Azure could remove this value from API but even this was agreed it won't happen in the near future. I'm tending to the second option, but before raising a PR it would be great to have some feedback if this is a acceptable way.

**Comment by dkisby** [comment]
Hi @wasfree

I had a look at #1, and this doesn't seem to work. Because although I can omit the auto-backup config from the deployment and rely on the user to configure the auto-backup as per documentation, then if they ever run the deployment again then they are faced with the following issue:
```
# azurerm_mssql_virtual_machine.db_vm must be replaced
-/+ resource "azurerm_mssql_virtual_machine" "db_vm" {
      ~ id                    = "vmId" -> (known after apply)
      - r_services_enabled    = false -> null

      - auto_backup { # forces replacement
          - encryption_enabled              = true -> null
          - retention_period_in_days        = 30 -> null
          - system_databases_backup_enabled = true -> null
        }

      + auto_patching {
          + day_of_week                            = "Saturday"
          + maintenance_window_duration_in_minutes = 60
          + maintenance_window_starting_hour       = 6
        }
    }
```
Because this forces replacement, it tries to destroy the azurerm_mssql_virtual_machine and redeploy, but it can't complete this process and reaches timeout during the destroy process. Leaving the SQL VM in a broken state.

#2 sounds like the answer to me.

**Comment by wasfree** [comment]
Hi @dkisby,

there has to be a misunderstanding in No. 1, the [pre-checking](https://github.com/terraform-providers/terraform-provider-azurerm/blob/50a50b3db89aa71b979c669d1e78bddcf737962d/azurerm/internal/services/mssql/mssql_virtual_machine_resource.go#L308) has to be disabled in terraform provider code not in terraform configuration. It won't work to configure this outside of terraform because `encryption_enabled` has a [default](https://github.com/terraform-providers/terraform-provider-azurerm/blob/50a50b3db89aa71b979c669d1e78bddcf737962d/azurerm/internal/services/mssql/mssql_virtual_machine_resource.go#L75) value of false which applies if auto_backup block is defined.

**Comment by dkisby** [comment]
Hi @wasfree 

Ah, my apologies, I misread No. 1.  Either way, I still think No. 2 sounds like the way forward, if pre-checking provides no value then deprecate it.

**Comment by jdelforno** [comment]
Atm, we're forced to split out the  deployment pipeline jobs to first, generate a secret and then use said secret when deploying backup encryption.

It's not a huge hassle but it's definitely inconvenient.

**Comment by PtrckGdwn** [comment]
Hi @wasfree,

I was wondering if the solution you outlined is something that can be implemented? It would make life quite a bit easier / cleaner if it is doable.


> Hi @dkisby,
> 
> seems this issue is related to a [pre-checking](https://github.com/terraform-providers/terraform-provider-azurerm/blob/50a50b3db89aa71b979c669d1e78bddcf737962d/azurerm/internal/services/mssql/mssql_virtual_machine_resource.go#L308) if `encryption_enabled` is true and `encryption_password` has a empty string. Unfortunately the `random_password` resource will provide the result only after apply, while the pre-checking expecting a non-empty string on a earlier stage.
> 
> There are few ways to resolve this issue:
> 
> 1. Remove this pre-checking and rely on documentation. So it's up to the user to provider a proper configuration.
> 2. We could deprecate `encryption_enabled` and only look for `encryption_password`. If `encryption_password` was set we also enable encryption, if not we disable it. This would also make the pre-checking obsolete.
> 
> To be honest `encryption_enabled` isn't really needed at all, if someone want to encrypt his backup and provided a `encryption_password` this should be enough. As a third way Azure could remove this value from API but even this was agreed it won't happen in the near future. I'm tending to the second option, but before raising a PR it would be great to have some feedback if this is a acceptable way.

**Comment by bel-from-nz** [comment]
Wasn't sure if I should start a new bug as this is related to the above.

If you add a lifecycle to ignore changes and deploy, then go into the Azure portal and set the auto backup to storage account on with encryption enabled, this pre-check is triggered on the next plan done.

```
resource "azurerm_mssql_virtual_machine" "sql_vm" {
  virtual_machine_id = module.vm_windows.resource_id
  sql_license_type   = "PAYG"

  lifecycle {
    ignore_changes = all
  }
}
```
Expected outcome: changes are ignored

Actual outcome: Error on plan

```
╷
│ Error: auto_backup: `encryption_password` is required when `encryption_enabled` is true
│
│   with azurerm_mssql_virtual_machine.sql_vm,
│   on main.tf line 90, in resource "azurerm_mssql_virtual_machine" "sql_vm":
│   90: resource "azurerm_mssql_virtual_machine" "sql_vm" {
│
╵
```

**Comment by kewalaka** [comment]
@wasfree this issue persists on azurerm 4.9 as well - is it worth updating the v2.x legacy label so it can be noted as an on-going problem.  Your option 2 looks reasonable to me.

**Comment by bel-from-nz** [comment]
> @wasfree this issue persists on azurerm 4.9 as well - is it worth updating the v2.x legacy label so it can be noted as an on-going problem. Your option 2 looks reasonable to me.

@rcskosir wasn't sure if the above had been seen. I am having issues with the azurerm_mssql_virtual_machine resource and was hoping for some update about the problem. Currently using azurerm 3.116.0 but tested this on 4.9.0 and it is still a problem.

**Comment by rcskosir** [comment]
:wave: Thanks for reaching out, unfortunately I do not have an update on this issue, but I can update the labels to mark it as still occurring in 3.x and 4.x. Any future work via the team or the community should end up linked here via a PR.

**Comment by bel-from-nz** [comment]
> 👋 Thanks for reaching out, unfortunately I do not have an update on this issue, but I can update the labels to mark it as still occurring in 3.x and 4.x. Any future work via the team or the community should end up linked here via a PR.

Thank you :)

**Comment by cgraf-spiria** [comment]
A workaround is to trick the pre-checking like this:
```
encryption_enabled   = length(random_password.encryptionpassword.result) > 1
encryption_password  = random_password.encryptionpassword.result
```
or add some text to the password: 

```
encryption_enabled   = true
encryption_password  = "x${random_password.encryptionpassword.result}"
```

But either way (and even when hardcoding a password) I get this error :( seems like another bug ?
```
 Azure API returned the following error:
│ 
│ Status: "Ext_AutomatedBackupError"
│ Code: ""
│ Message: "Error: 'The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.'"
│ Activity Id: ""
```

**Comment by cgraf-spiria** [comment]
I found the details of my Base-64 error in the logs of `C:\WindowsAzure\Logs\Plugins\Microsoft.SqlServer.Management.SqlIaaSAgent`
```
System.Exception: Configuring auto backup failed     
---> System.FormatException: The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.        at System.Convert.FromBase64String(String s)        
at Azure.Storage.StorageSharedKeyCredential..ctor(String accountName, String accountKey)        
at Microsoft.SqlServer.Management.AutoBackup.ManagedBackupV2Helpers.CreateSASKeyAndPolicy(String storageName, String storageAccountAccessKey, String containName, String storageUrl)        
at Microsoft.SqlServer.Management.AutoBackup.ManagedBackupQueriesV2.CreateManagedBackupCredential(ManagedBackupSettings settings, Boolean forceRecreate)        
at Microsoft.SqlServer.Management.AutoBackup.ManagedBackupQueriesV2.UpdateManagedBackupConfigurations(ManagedBackupSettings settings, String& errorMessage)        
at Microsoft.SqlServer.Management.AutoBackup.ManagedBackupManagement.UpdateConfigurations(SecureString protectedSettingsCertificateThumbprint, SecureString protectedSettings, PublicSettings publicSettings)        
--- End of inner exception stack trace --
```

Turns out I was using `primary_connection_string` instead of `primary_access_key` when assigning `storage_account_access_key`🤦


As for the bug in this issue, like mentioned above https://github.com/hashicorp/terraform-provider-azurerm/issues/12811#issuecomment-2469485729, the validation will always fail when `ignore_changes` is set. 
Which sadly makes the `azurerm_mssql_virtual_machine` resource unusable if you want to configure automated backups manually.

I created a PR to remove the validations, deprecate the `encryption_enabled` attribute and add a validation for the access key
